coral-schema/src/main/java/com/linkedin/coral/schema/avro/SchemaUtilities.java

```diff
@@ -436,46 +445,62 @@ private static boolean isUnionSchemaCompatible(@Nonnull Schema leftSchema, @Nonn
       case INT:
       case LONG:
       case STRING:
-      case NULL:
-        return leftSchema.getType() == rightSchema.getType();
+        outputSchema = leftSchema;
+        break;
       case FIXED:
-        boolean isSameType = leftSchema.getType() == rightSchema.getType();
-        boolean isSameNamespace = Objects.equals(leftSchema.getNamespace(), rightSchema.getNamespace());
-
-        return isSameType && (!strictMode || isSameNamespace);
+        compatible &= isSameNamespace(leftSchema, rightSchema, strictMode);
+        outputSchema = leftSchema;
+        break;
       case ENUM:
-        boolean isSameEnumType = (leftSchema.getType() == rightSchema.getType());
         boolean isSameSymbolSize = (leftSchema.getEnumSymbols().size() == rightSchema.getEnumSymbols().size());
```

Comment:
There is an issue if we don't check the type of `rightSchema` before calling `rightSchema.getEnumSymbols()`, which also happens in `RECORD`, `MAP`, `ARRAY`, maybe we need to throw the exception first if `compatible == false` before `switch`.

Reply:
Good point. Refactored. Please take another look.